File: Piko/Sources/SFTriangleNode.m
//
//  SFTriangleNode.m
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

#import "SFTriangleNode.h"

@interface SFTriangleNode ()

@property (nonatomic, assign) double phase;

@end

@implementation SFTriangleNode

- (instancetype)init
{
    SFInput *frequency = [[SFInput alloc] initWithName:@"frequency"];
    SFOutput *output = [[SFOutput alloc] initWithName:@"output"];
    self = [super initWithInputs:@[frequency] outputs:@[output]];
    if (self != nil) {
        _frequency = frequency;
        _output = output;
    }
    return self;
}

- (BOOL)rendersConstantFrame
{
    return NO;
}

- (BOOL)renderForPlayer:(SFPlaybackManager *)player
                  index:(UInt64)index
                  count:(NSInteger)count
                     to:(double *)buffer
{
    SFNode *frequencyNode = _frequency.source.owner;

    if (frequencyNode == nil) {
        return NO;
    }

    if (![frequencyNode rendersConstantFrame]) {
// TODO: 未対応
        return NO;
    }

    double freq = 0.0;

    if (![frequencyNode renderForPlayer:player index:index count:1 to:&freq]) {
        return NO;
    }

    double step = freq / player.sampleRate;

    for (NSInteger i = 0; i < count; i++) {
        buffer[i] = (_phase < 0.25) ? _phase * 4.0 :
                    (_phase < 0.75) ? (0.50 - _phase) * 4.0 :
                                     (_phase - 1.0) * 4.0;
        _phase += step;
        while (1.0 <= _phase) {
            _phase -= 1.0;
        }
    }
    return YES;
}

@end

File: Piko/Sources/SFTriangleNode.swift
//
//  SFTriangleNode.swift
//  Piko
//
//  Created by Satoshi Muraki on 8/14/14.
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

import Foundation

@objc class SFTriangleNode: SFNode {

    let frequency: SFInput
    let output: SFOutput

    var phase: Double = 0.0

    init() {
        frequency = SFInput(name: "frequency")
        output = SFOutput(name: "output")
        super.init(inputs: [frequency], outputs: [output])
    }

    override func renderForPlayer(player: SFPlaybackManager,
                                   index: UInt64,
                                   count: Int,
                               to buffer: UnsafeMutablePointer<Double>) -> Bool {

        let frequencyNode = frequency.source?.owner!

        if frequencyNode == nil {
            return false
        }

        if frequencyNode!.rendersConstantFrame() == false {
            return false
        }

        var freq: Double = 0.0

        if player.render(frequencyNode, index: index, count: 1, to: &freq) == false {
            return false
        }

        let step = freq / player.sampleRate

        for i in 0..<count {
            buffer[i] = (phase < 0.25) ? phase * 4.0 :
                        (phase < 0.75) ? (0.50 - phase) * 4.0 :
                                         (phase - 1.0) * 4.0
            phase += step;
            while 1.0 <= phase {
                phase -= 1.0
            }
        }
        return true
    }
}
